syntax = "proto3";
package test;

import "validate.proto";

// Message type codes
enum MessageType {
  MSG_UNKNOWN = 0;
  MSG_PING = 1;
  MSG_PONG = 2;
  MSG_REQUEST = 3;
  MSG_RESPONSE = 4;
  MSG_ERROR = 5;
  MSG_NOTIFICATION = 6;
}

// Ping message
message Ping {
  uint64 timestamp = 1 [(validate.rules).uint64.gt = 0];
  uint32 sequence = 2;
}

// Pong message  
message Pong {
  uint64 timestamp = 1 [(validate.rules).uint64.gt = 0];
  uint32 sequence = 2;
  uint32 latency_ms = 3;
}

// Request message
message Request {
  string method = 1 [(validate.rules).string.min_len = 1];
  bytes payload = 2 [(validate.rules).bytes.max_len = 1024];
  uint32 request_id = 3;
}

// Response message
message Response {
  uint32 request_id = 1;
  int32 status_code = 2 [(validate.rules).int32 = {gte: 0, lt: 600}];
  bytes payload = 3 [(validate.rules).bytes.max_len = 4096];
}

// Error message
message Error {
  uint32 error_code = 1 [(validate.rules).uint32.gt = 0];
  string message = 2 [(validate.rules).string.min_len = 1];
  string details = 3;
}

// Notification message
message Notification {
  string event_type = 1 [(validate.rules).string.min_len = 1];
  bytes data = 2;
  uint64 timestamp = 3;
}

// Main envelope message with header
message Envelope {
  uint32 version = 1 [(validate.rules).uint32 = {gte: 1, lte: 10}];
  MessageType msg_type = 2 [(validate.rules).enum.defined_only = true];
  uint64 correlation_id = 3;
  
  oneof message {
    Ping ping = 10;
    Pong pong = 11;
    Request request = 12;
    Response response = 13;
    Error error = 14;
    Notification notification = 15;
  }
}

// RPC service definitions
service TestService {
  rpc SendPing(Ping) returns (Pong);
  rpc ProcessRequest(Request) returns (Response);
  rpc Subscribe(Notification) returns (Notification);
}
